compiler/rustc_privacy/src/lib.rs

```diff
@@ -241,6 +256,25 @@ where
             ty.super_visit_with(self)
         }
     }
+
+    fn visit_const(&mut self, c: &'tcx Const<'tcx>) -> ControlFlow<Self::BreakTy> {
+        self.visit_ty(c.ty)?;
+        let tcx = self.def_id_visitor.tcx();
+        use rustc_middle::mir::abstract_const::Node as ACNode;
+        use rustc_middle::ty::subst::Subst;
+        if let Ok(Some(ct)) = AbstractConst::from_const(tcx, c) {
```

Comment:
Visiting the type (`self.visit_ty(c.ty)` above) is certainly necessary because by leaking a constant you are leaking its type as well.

The part starting from this line and below doesn't make much sense to me.
Requiring intermediate constants or functions to be public gives no guarantees about them being usable by wrappers to propagate requirements upwards (they can easily be public but unnameable).
Marking them as reachable is also not necessary, since they are "ephemeral" and only the evaluation result is leaked.

Reply:
the current intent is to copy the behavior of trait bounds as much as possible, so this PR should cause us to end up with the same behavior as
```rust
trait Bound {}

pub fn foo<T: Bound>() {} // err

mod hidden {
    pub trait Bound {}
}

pub fn bar<T: hidden::Bound>() {} // ok
```
While this isn't something I am too involved in, I do think that being consistent here is valuable.

Reply:
I won't argue much about this since it's a conservative check which can always be relaxed later, but the analogy here is pretty superficial.